api/v1/stats/service.go

```diff
@@ -0,0 +1,185 @@
+	aggAppendSum(s, c, "server stats stream append_in_record")
+}
+
+func (s *Service) GetSendOutRecords(c *gin.Context) {
+	aggAppendSum(s, c, "server stats subscription send_out_records")
+}
+
+func (s *Service) GetAppendTotal(c *gin.Context) {
+	aggAppendSum(s, c, "server stats stream_counter append_total")
+}
+
+func (s *Service) GetAppendFailed(c *gin.Context) {
+	aggAppendSum(s, c, "server stats stream_counter append_failed")
+}
+
+func (s *Service) GetServerAppendRequestLatency(c *gin.Context) {
+	aggSum(s, c, "server stats server_histogram append_request_latency")
+}
+
+func (s *Service) GetServerAppendLatency(c *gin.Context) {
+	aggSum(s, c, "server stats server_histogram append_latency")
+}
+
+func adminError(err error) errorno.ErrorResponse {
+	return errorno.NewErrorResponse(errorno.ADMIN_GET_STATUS_ERROR, err)
+}
+
+func getFromCluster(s *Service, cmd string) ([]model.TableResult, error) {
+	info, err := s.client.GetServerInfo()
+	if err != nil {
+		return nil, fmt.Errorf("GetServerInfo error: %v", err)
+	}
+
+	allStatsRaw := []model.TableResult{}
+
+	for _, addr := range info {
+		respTable, err := s.client.GetStatsFromServer(addr, cmd)
+		if err != nil {
+			return nil, fmt.Errorf("AdminRequest error with %v: %v", cmd, err)
+		}
+
+		stats := model.TableResult{
+			Headers: respTable.Headers,
+			Value:   make([]map[string]string, 0, len(respTable.Rows)),
+		}
+
+		for _, row := range respTable.Rows {
+			mp := make(map[string]string, len(row))
+			for idx, val := range row {
+				mp[respTable.Headers[idx]] = val
+			}
+			stats.Value = append(stats.Value, mp)
+		}
+
+		allStatsRaw = append(allStatsRaw, stats)
+	}
+
+	return allStatsRaw, nil
+}
+
+func aggAppendSum(s *Service, c *gin.Context, cmd string) {
+	allStatsRaw, err := getFromCluster(s, cmd)
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusInternalServerError, adminError(err))
+		return
+	}
+
+	if len(allStatsRaw) == 0 {
+		c.AbortWithStatusJSON(http.StatusInternalServerError, adminError(fmt.Errorf("got stats of length 0")))
+		return
+	} else {
+		headers := allStatsRaw[0].Headers
+
+		allStats := model.TableResult{
+			Headers: headers,
+			Value:   []map[string]string{},
+		}
+
+		// The `allStatsRaw` contains stats tables that got from the cluster per server, which each is of the form:
+		// | main_key | type_0 stats | type_1 stats | ...
+		// If the main_key already contains in the container which is prepared for return the result, add to exist stats
+		// for each value.  This can happen when a stream, subscription or something else if exists, which transfer to
+		// be handled by another server. If main_key does not already contain in, just append the value col to the
+		// table result.
+		for _, stats := range allStatsRaw {
+			for _, val := range stats.Value {
+				mainKey := val[headers[0]]
+
+				ok := false
+				ix := -1
+				for i, cur := range allStats.Value {
+					if cur[mainKey] == val[mainKey] {
+						ok = true
+						ix = i
+					}
+				}
```

Comment:
If you just want to find the idx in allStats.Value which `allStats.Value[idx][mainKey] == val[mainKey]`, why you don't break the loop when you find the idx?

Reply:
thx! fixed in [7bc88a4](https://github.com/hstreamdb/http-services/pull/12/commits/7bc88a4c758dc7bf50dbfeff1d1170155097aa93)